bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1536 @@
+// Rust Bitcoin Library - Written by the rust-bitcoin developers.
+// SPDX-License-Identifier: CC0-1.0
+
+//! Proof-of-work related integer types.
+//!
+//! Provides the [`Work`] and [`Target`] types that are use in proof-of-work calculations. The
+//! functions here are designed to be fast, by that we mean it is safe to use them to check headers.
+//!
+
+use core::fmt::{self, LowerHex, UpperHex};
+use core::ops::{Add, Div, Mul, Not, Rem, Shl, Shr, Sub};
+
+use crate::consensus::encode::{self, Decodable, Encodable};
+#[cfg(doc)]
+use crate::consensus::Params;
+use crate::hash_types::BlockHash;
+use crate::io::{self, Read, Write};
+
+/// Implements $int * $ty. Requires `u64::from($int)`.
+macro_rules! impl_int_mul {
+    ($ty:ident, $($int:ident),+ $(,)?) => {
+        $(
+            impl Mul<$ty> for $int {
+
+                type Output = $ty;
+                #[inline]
+                fn mul(self, rhs: $ty) -> $ty { $ty(self.mul(rhs.0)) }
+            }
+        )+
+    };
+}
+
+/// Implement traits and methods shared by `Target` and `Work`.
+macro_rules! do_impl {
+    ($ty:ident) => {
+        impl $ty {
+            /// Creates `Self` from a big-endian byte array.
+            #[inline]
+            pub fn from_be_bytes(bytes: [u8; 32]) -> $ty { $ty(U256::from_be_bytes(bytes)) }
+
+            /// Creates `Self` from a little-endian byte array.
+            #[inline]
+            pub fn from_le_bytes(bytes: [u8; 32]) -> $ty { $ty(U256::from_le_bytes(bytes)) }
+
+            /// Converts `self` to a big-endian byte array.
+            #[inline]
+            pub fn to_be_bytes(self) -> [u8; 32] { self.0.to_be_bytes() }
+
+            /// Converts `self` to a little-endian byte array.
+            #[inline]
+            pub fn to_le_bytes(self) -> [u8; 32] { self.0.to_le_bytes() }
+        }
+
+        impl fmt::Display for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
```

Comment:
This formats `Work` and `Target` as decimals. Did we really intend that?

Reply:
I can't remember what made me do this but it was intentional at the time because I remember writing the `fmt_decimal` function. Whether or not we want to do this I don't know right at the moment (my brain is in holiday mode I don't really know whats going on, just checking my PRs to push trivial things forward).

Reply:
OK, I guess let's go with it. They are numbers, not hashes so formatting them as hex is not obvious and there's no better way to explicitly format as decimals.

Reply:
I also don't remember this discussion, but my guess is that because Bitcoin Core logs these values as decimals, we decided to go with that

Reply:
Oh, that's great to hear we're not the only strange ones doing that. :)

Also :tada: for merging this!

Reply:
I think Core also logs the log2 value as a float, which is an interesting idea but probably needs its own discussion and bikeshedding if we wanted to do it in `Display`. (Maybe it should be the alternate display? I dunno.)

Reply:
Maybe just provide conversion to `f64`?

Reply:
I like that (called `.log2()` or something, not `.into` ;)).

Reply:
`log2` sounds good.